frontend/public/module/k8s/autocomplete.ts

```diff
@@ -1,223 +0,0 @@
-import { IEditSession, Editor, Position } from 'brace';
-import { Map as ImmutableMap } from 'immutable';
-import * as _ from 'lodash-es';
-
-import { ServiceAccountModel, SecretModel, ServiceModel, ConfigMapModel, AlertmanagerModel } from '../../models';
-import { getStoredSwagger, K8sKind, K8sResourceKind, SwaggerDefinitions } from '../../module/k8s';
-import { k8sListPartialMetadata } from '../../module/k8s/resource';
-import { getActiveNamespace } from '../../actions/ui';
-
-export type AceSnippet = {
-  content: string;
-  name?: string;
-  guard?: string;
-  trigger?: string;
-  endTrigger?: string;
-  endGuard?: string;
-  tabTrigger?: string;
-};
-
-export const snippets = ImmutableMap<string, string>()
-  .set('metadata', `
-    metadata:
-      name: \${1}
-      namespace: \${2}
-      labels: 
-        \${3}
-      annotations:
-        \${4}
-  `)
-  .set('selector', `
-    selector:
-      matchLabels:
-        \${1}
-  `)
-  .set('containers', `
-    containers:
-      - name: \${1}
-        image: \${2}
-        ports:
-          - containerPort: \${3}
-        volumeMounts:
-          - name: \${4}
-            mountPath: \${5}
-        command:
-          - \${6}
-  `)
-  .set('volumeClaimTemplates', `
-    volumeClaimTemplates:
-      - metadata:
-          name: \${1}
-        spec:
-          accessModes:
-            - ReadWriteOnce
-          storageClassName: \${2}
-          resources:
-            requests:
-              storage: 1Gi
-  `)
-  .set('rules', `
-    rules:
-      - http:
-          paths:
-            - path: /\${1}
-              backend:
-                serviceName: \${2}
-                servicePort: 80
-  `)
-  .set('ports', `
-    ports:
-      - protocol: TCP
-        port: 80
-        targetPort: \${1}
-  `)
-  .set('parameters', `
-    parameters:
-      - name: \${1}
-        value: \${2}
-        description: \${3}
-  `)
-  .set('triggers', `
-    triggers:
-      - imageChange:
-          from:
-            kind: ImageStreamTag
-            name: \${1}
-        type: ImageChange
-  `)
-  .map((content, name) => ({name, content: content.split('\n').slice(1, content.split('\n').length - 1).map(s => s.substring(4)).join('\n')} as AceSnippet));
-
-/**
- * Defines a bunch of possibilities for property value completion, defined in the OpenAPI/Swagger spec.
- */
-export const simpleCompletions = ImmutableMap<string, string[]>()
```

Comment:
As far as I can tell, we are losing features here. I tried to autocomplete `imagePullPolicy`, and got nothing. Whereas our existing implementation gives us `Always`, `Never`, and `IfNotPresent`. The fact is, the Kubernetes OpenAPI definitions are not that great. The valid values for some properties are often just stated in the description rather than proper `oneOf`. I also didn't get cluster-aware autocomplete for available `Secrets` or anything else. I'm open to contributing to the language server for these features, but right now it's actually worse.

Reply:
OK, I have to disagree with this assessment. It's way, way better because it's context sensitive. If you press control space, you only are prompted for the properties that are valid in that stanza. You also get hover help and it will fill in required properties automatically.

Note that there is some room for improvement in our translation of OpenAPI doc -> JSON schema, so there are some gaps. I'm working on making this better.

Reply:
The other way this is better is that it has full completion for all types without our having to code anything manually. If a new type is added or something changes in a future k8s version, it just works.

Reply:
> I'm open to contributing to the language server for these features

These aren't problems with the language server, rather they're problems with the kubernetes OpenAPI.

> As far as I can tell, we are losing features here

Perhaps losing a couple but gaining a lot. Also, its probably possible to augment the schema so that you can still have auto completion results such as `Always`, `Never`, and `IfNotPresent`

> the Kubernetes OpenAPI definitions are not that great

Maybe this is the perfect time to lobby for improvements? The better the OpenAPI the better the support in Openshift Console would be